### PR TITLE
sanitycheck: coverage improvements

### PR DESCRIPTION
**Pull request by mniestroj**
Add `--coverage-basedir BASEDIR` argument, which can be used to specify
source code base directory other than default Zephyr root directory.

Match any directory in ignored path hierarchy, similar how file is matched
currently. That way sanitycheck can be executed from outside of Zephyr
source code directory and still respect ignored directories list.

Both improvements are very helpful for Zephyr based application code 
unit testing.

### CODE REVIEW

**Comment by tribal-tec** [comment]
Is this planned to be merged for the 2.4 release? Just tested myself, and it does what I was looking for! Would be really appreciated to see this in the upcoming version.

**Review by nashif — APPROVED** [review]

**Comment by tribal-tec** [comment]
Thanks for following up on this. I just noticed however if the `coverage-basedir` is outside of the zephyrproject directory, it filters everything and doesn't include my app code. I would have to check on this tomorrow.

**Comment by tribal-tec** [comment]
Ok, my bad. I was expecting coverage from files within the `tests` subfolder to be shown, but they are of course filtered as well. All good then when using `gcovr`. However, with `lcov`, the filtering of Zephyr root files is not working though.

File: scripts/sanity_chk/sanitylib.py [hunk 1]
```diff
@@ -3350,7 +3350,7 @@ def add_ignore_file(self, pattern):
         self.ignores.append('*' + pattern + '*')
 
     def add_ignore_directory(self, pattern):
-        self.ignores.append(pattern + '/*')
+        self.ignores.append('*/' + pattern + '/*')
```
Comment (PerMac) [inline scripts/sanity_chk/sanitylib.py:3353]: In the option below you have `(".*/" + pattern + '/.*')`. Here it is `('*/' + pattern + '/*')` so it looks like `.` might be missing before `*`. Can this be the reason why @tribal-tec is observing problems with `lcov`?

Comment (mniestroj) [reply scripts/sanity_chk/sanitylib.py:3353]: `lcov` and `gcov` have both different syntax for specifying ignore patterns. One requires `*/` and the other requires `.*/` to get the same effect.

**Comment by mniestroj** [comment]
>Ok, my bad. I was expecting coverage from files within the tests subfolder to be shown, but they are of course filtered as well. All good then when using gcovr. However, with lcov, the filtering of Zephyr root files is not working though.

Could you elaborate what is not working? What is your use case, what do you expect and what happens instead?

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by tribal-tec** [comment]
> > Ok, my bad. I was expecting coverage from files within the tests subfolder to be shown, but they are of course filtered as well. All good then when using gcovr. However, with lcov, the filtering of Zephyr root files is not working though.
> 
> Could you elaborate what is not working? What is your use case, what do you expect and what happens instead?

It was a mistake on my side, I just had code I wanted coverage result for in the tests subfolder, but this is not something you would have in a real application. The tests folder is correctly filtered from the coverage results, so all good then 😄

**Review by PerMac — APPROVED** [review]

**Pull request merged**